nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/opvalidation/ReductionOpValidation.java

```diff
@@ -1058,4 +1060,168 @@ public void testReductionsBackwards() {
             sd.execBackwards(Collections.emptyMap());
         }
     }
+
+    @Test
+    public void testDotProductAttention(){
+        final INDArray keys = Nd4j.rand(new int[]{10, 4, 3});
+        final INDArray values = Nd4j.rand(new int[]{10, 4, 3});
+        final INDArray query = Nd4j.rand(new int[]{10, 4, 1});
+
+        final INDArray exec = Nd4j.matmul(keys, query, true, false, false)
+                .divi(Math.sqrt(keys.size(1)));
+        Nd4j.exec(new SoftMax(exec, exec, 1));
+        final INDArray finalOut = Nd4j.matmul(values, exec).norm1();
+
+        SameDiff sd = SameDiff.create();
+        SDVariable sdQ = sd.var("q", query);
+        SDVariable sdK = sd.var("k", keys);
+        SDVariable sdV = sd.var("v", values);
+
+        SDVariable t = sd.nn.dotProductAttention(sdQ, sdK, sdV, null, true);
+        t.norm1("out");
+
+        String err = OpValidation.validate(new TestCase(sd)
+                    .expectedOutput("out", finalOut)
+                    .gradientCheck(true));
+        assertNull(err, err);
```

Comment:
Really minor point: just assertNull(err) is fine. It actually prints the object if it fails, even without the string set as the message.
(And yes, I know a ton of my tests do have assertNull(err,err) but I'm cleaning those as I see them)